### ISSUE TITLE
Exceptions coming from env['action_dispatch.exception'] are not caught by Raven::Rack

### ISSUE THREAD

**Issue by Ferdy89** [getsentry/raven-ruby#720]
Hi, we're currently using Sentry Raven 2.5.3 on a Rails 4.2.8 application with pretty much no configuration for the gem other than the automatic one it provides. We're noticing that some exceptions are not caught by the `Raven::Rack` middleware.

Here's how our middleware stack looks like in production:
```
use Raven::Rack
# ...
use ActionDispatch::ShowExceptions
# ...
run Rails::Application.routes
```

In production, `Rails.configuration.action_dispatch.show_exceptions` is set to `true` by default, which means [the `ActionDispatch::ShowExceptions` middleware catches and stops any exceptions](https://github.com/rails/rails/blob/v4.2.8/actionpack/lib/action_dispatch/middleware/show_exceptions.rb#L35). It is graceful enough to [leave the exception in the `env`](https://github.com/rails/rails/blob/v4.2.8/actionpack/lib/action_dispatch/middleware/show_exceptions.rb#L44) under the name of `'action_dispatch.exception'`, though.

However, Raven then [fails to look at that part of the `env`, even though it does look at some parts of the `env` for "hidden" exceptions](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rack.rb#L58).

We wonder if it's that we're using the gem in a way that it's not designed for or if it's that the gem should be also looking at that part of the `env`. Could you tell us which direction to go through? We'd be happy to submit better reproduction steps or a patch for the gem.

Thank you!

**Comment by nateberkopec** [comment]
> some exceptions

Can you expand on what you mean by that?

**Comment by Ferdy89** [comment]
Sorry about the ambiguity. We have our application to rescue errors at the `ApplicationController` level so we explicitly handle most errors, like:
```ruby
class ApplicationController < ActionController::Base
  rescue_from StandardError do |exception|
    Raven::Rack.capture_exception(exception)
  end
end
```
In this particular case, the error was being generated in the routes, so the controller wasn't able to catch it. That's why I said "some" exceptions, I should have said all exceptions.

Now, I went ahead and created a brand new Rails app and followed these steps for reproduction:
* Create a controller and an action which just raises an exception.
* Add the `sentry-raven` gem to the project.
* Add a key in `config/secrets.yml` to be able to boot in production mode.
* Using `bundle open sentry-raven`, I threw a `puts error if error` [right at the end of the `Raven::Rack` middleware](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rack.rb#L60).
* I booted up the app and hit the homepage and saw the 500 page, but nothing in the console.
* I then did `bundle open sentry-raven` again and changed [the line where it catches errors from other frameworks](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rack.rb#L58) and added `|| env['action_dispatch.exception']`.
* I hit the homepage again and confirmed that in console the error is captured this time.

I hope that helps reproducing what we're experiencing. Thanks!

**Comment by nateberkopec** [comment]
OK. I'll look at this later, but the reason we don't look in action_dispatch.exception is because we [override render_exception.](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rails/overrides/debug_exceptions_catcher.rb). See also [here](https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rails.rb#L45).

**Comment by cmoylan** [comment]
@Ferdy89 Following your instructions, I added an additional `puts exception` right here - https://github.com/getsentry/raven-ruby/blob/master/lib/raven/integrations/rails/overrides/debug_exceptions_catcher.rb#L6. When I hit the homepage, I see the output from this statement in the logs. I think `render_exception` is being successfully overridden by Raven and your issue is coming from somewhere else, possibly configuration.

**Comment by nateberkopec** [comment]
@Ferdy89 IIRC our rack middlewares sometimes interfact poorly with custom exception applications, if you have one of those configured.

**Comment by cmoylan** [comment]
@nateberkopec This issue can be closed, The true cause of our issue is described and fixed in #722
